libtcmu.c

```diff
@@ -173,6 +197,433 @@ static struct tcmulib_handler *find_handler(struct tcmulib_context *ctx,
+{
+	int ret;
+	struct tcmu_io_queue *io_wq = &dev->work_queue;
+
+	list_head_init(&io_wq->io_queue);
+
+	ret = pthread_mutex_init(&io_wq->io_lock, NULL);
+	if (ret < 0) {
+		goto out;
+	}
+	ret = pthread_cond_init(&io_wq->io_cond, NULL);
+	if (ret < 0) {
+		goto cleanup_lock;
+	}
+
+	// TODO: >1 worker threads (per device via config)
+	ret = pthread_create(&io_wq->io_wq_thread, NULL, io_work_queue, dev);
+	if (ret < 0) {
+		goto cleanup_cond;
+	}
+
+	return 0;
+
+cleanup_cond:
+	pthread_cond_destroy(&io_wq->io_cond);
+cleanup_lock:
+	pthread_mutex_destroy(&io_wq->io_lock);
+out:
+	return ret;
+}
+
+static void cleanup_io_work_queue(struct tcmu_device *dev,
+				  bool cancel)
+{
+	int ret;
+	struct tcmu_io_entry *curr, *next;
+	struct tcmu_io_queue *io_wq = &dev->work_queue;
+
+	if (cancel) {
+		cancel_thread(io_wq->io_wq_thread);
+	}
+
+	list_for_each_safe(&io_wq->io_queue, curr, next, entry) {
+		list_del(&curr->entry);
+		free(curr);
+	}
+
+	ret = pthread_mutex_destroy(&io_wq->io_lock);
+	if (ret != 0) {
+		tcmu_err("failed to destroy io workqueue lock\n");
+	}
+
+	ret = pthread_cond_destroy(&io_wq->io_cond);
+	if (ret != 0) {
+		tcmu_err("failed to destroy io workqueue cond\n");
+	}
+}
+
+static int setup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+	aio_track->tracked_aio_ops = 0;
+	ret = pthread_spin_init(&aio_track->track_lock, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return 0;
+}
+
+static void cleanup_aio_tracking(struct tcmu_device *dev)
+{
+	int ret;
+	struct tcmu_track_aio *aio_track = &dev->track_queue;
+
+        assert(aio_track->tracked_aio_ops == 0);
+
+	ret = pthread_spin_destroy(&aio_track->track_lock);
+	if (ret < 0) {
+		tcmu_err("failes to destroy track lock\n");
+	}
+}
+
+static void async_call_command(struct tcmu_device *dev,
+			       struct tcmulib_cmd *cmd)
+{
+	int ret;
+	struct tcmulib_handler *handler = tcmu_get_dev_handler(dev);
+	struct tcmur_handler *r_handler = handler->hm_private;
+
+	if (r_handler->aio_supported) {
+		ret = invokecmd(dev, cmd);
+	} else {
+		ret = aio_schedule(dev, cmd);
+	}
+
+	/*
+	 * command (processing) completion is done when one of the
+	 * following scenario occurs:
+	 *  - synchronous handler:
+	 *	only if aio_schedule() returns an error
+	 *  - asynchronous handler:
+	 *	on an error or if the command was not asynchronously
+	 *	handled (see generic_handle_cmd(), non store callouts)
+	 */
+	if (ret != TCMU_ASYNC_HANDLED) {
+		tcmulib_command_complete(dev, cmd, ret);
+		tcmulib_processing_complete(dev);
+	}
+}
+
+static void *tcmu_cmdproc_thread(void *arg)
+{
+	struct tcmu_device *dev = arg;
+	struct pollfd pfd;
+
+	pthread_cleanup_push(cmdproc_thread_cleanup, dev);
+
+	while (1) {
+		struct tcmulib_cmd *cmd;
+
+		tcmulib_processing_start(dev);
+
+		while ((cmd = tcmulib_get_next_command(dev)) != NULL) {
+			if (tcmu_get_log_level() == TCMU_LOG_DEBUG)
+				tcmu_cdb_debug_info(cmd);
+
+			/* invoke async command */
```

Comment:
I would just drop the comment or rename the function. They differ by just one word (invoke vs call) :)